engine: let maxKeys=0 scan unboundedly

tbg opened the pull request:
First two commits are https://github.com/cockroachdb/cockroach/pull/45008, please ignore them here.

----

For historical reasons, specifying maxKeys=0 meant "return no results".
Following the Go koan of "making the zero value useful", zero should
correspond to "no limit". In particular, maxKeys can move to
MVCCScanOptions, where it belongs.

This PR makes that change in two steps. First, it adds MaxKeys to
MVCCScanOptions and updates all callers to use it there. To keep
the review fun(ish), this first commit does not yet drop the original
maxKeys parameter that threads itself through large chunks of the
codebase. Instead, there are assertions that make sure the old one
is essentially unused: all callers will supply `-1` (=ignore me) or
`MaxInt64` (=ignore me). The next commit then drops these parameters
altogether.

Note that no external contracts were changed here, i.e., there isn't a
migration concern.

Release note: None

tbg committed:
storage: remove outdated maxKeys param from MVCCScan

This drops the extra parameter left over from the previous commit.

Release note: None

tbg committed:
storage: let maxKeys=0 scan unboundedly

For historical reasons, specifying maxKeys=0 meant "return no results".
Following the Go koan of "making the zero value useful", zero should
correspond to "no limit". In particular, maxKeys can move to
MVCCScanOptions, where it belongs.

This PR makes the first step in that direction. It adds MaxKeys to
MVCCScanOptions and updates all callers to use it there. To keep
the review fun(ish), this first commit does not yet drop the original
maxKeys parameter that threads itself through large chunks of the
codebase. Instead, there are assertions that make sure the old one
is essentially unused: all callers will supply `-1` (=ignore me) or
`MaxInt64` (=ignore me). The next commit then drops these parameters
altogether.

Note that no external contracts were changed here, i.e., there isn't a
migration concern.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45069)
<!-- Reviewable:end -->

tbg committed:
batcheval: remove CommandArgs.MaxKeys

It was already replaced by MaxSpanRequestKeys (sitting on the batch
header).

Release note: None

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> save for some minor points. Only reviewed the storage parts though.

> Reviewed 8 of 8 files at r1, 2 of 2 files at r2, 13 of 13 files at r3, 15 of 15 files at r4, 11 of 11 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-:-M0-nJhY7_yKxMk1Lr0I:b9adq89)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/storage/batcheval/cmd_resolve_intent_test.go, line 218 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-M0-nprdCotB2rmWmOWn:-M0-nprdCotB2rmWmOWo:be34azy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f45759dbf30a980f3c2c905059c7ed31052497a2/pkg/storage/batcheval/cmd_resolve_intent_test.go#L218)):*
> ```Go
> 			declareKeysResolveIntentRange(&desc, h, &rir, &spans)
> 
> 			h := h
> ```

unnecessary?
___
*[pkg/storage/engine/mvcc.go, line 2339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-M0-ntsP5tOTG3a2o34A:-M0-ntsQ3EUNRJm_8DfF:bnc5fht) ([raw file](https://github.com/cockroachdb/cockroach/blob/f45759dbf30a980f3c2c905059c7ed31052497a2/pkg/storage/engine/mvcc.go#L2339)):*
> ```Go
> 		end:              endKey,
> 		ts:               timestamp,
> 		maxKeys:          opts.MaxKeys,
> ```

Since maxKeys will never be a negative number, may as well emphasize that by changing its type to a `uint64` ? The same can be done with `max_keys_` in `mvcc.h`. Just a minor suggestion; what we have currently works too.


<!-- Sent from Reviewable.io -->

tbg committed:
engine: test maxKeys==-1

This case is used in practice when batch evaluation has already
saturated MaxSpanRequestKeys. In this case, we still want to
evaluate the remaining scans (but without returning any rows).
This is implemented with early returns, so nothing interesting
happens at the MVCC level, but it's still good to test this.

While I was there, I added a bit of general testing for MaxKeys
which is present elsewhere, but not in this nice datadriven format.

Release note: None

tbg reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-:-M0qX0gN1KdHTsf-o3gK:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @nvanbenschoten)

___
*[pkg/storage/batcheval/cmd_resolve_intent_test.go, line 218 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-M0-nprdCotB2rmWmOWn:-M0qRw4F7rHbDWYkOAWJ:b8a8f7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f45759dbf30a980f3c2c905059c7ed31052497a2/pkg/storage/batcheval/cmd_resolve_intent_test.go#L218)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

unnecessary?
</blockquote></details>

You're right, done.
___
*[pkg/storage/engine/mvcc.go, line 2339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45069#-M0-ntsP5tOTG3a2o34A:-M0qWIse0LtUYnpOCpja:b-ttw2d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f45759dbf30a980f3c2c905059c7ed31052497a2/pkg/storage/engine/mvcc.go#L2339)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Since maxKeys will never be a negative number, may as well emphasize that by changing its type to a `uint64` ? The same can be done with `max_keys_` in `mvcc.h`. Just a minor suggestion; what we have currently works too.
</blockquote></details>

It actually will sometimes be a negative number (see the -1 commit I added; -1 is basically what 0 was before this PR). I added some tests too.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=itsbilal
